File: src/main/java/com/paragon/client/systems/module/impl/render/Tracers.java
package com.paragon.client.systems.module.impl.render;

import com.paragon.api.util.entity.EntityUtil;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.client.systems.module.Module;
import com.paragon.client.systems.module.Category;
import com.paragon.client.systems.module.setting.Setting;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityEnderCrystal;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraftforge.client.event.RenderWorldLastEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

import java.awt.*;

/**
 * @author Wolfsurge
 */
public class Tracers extends Module {

    public static Tracers INSTANCE;

    public static Setting<Boolean> passive = new Setting<>("Passives", true)
            .setDescription("Draws lines to passive entities");

    public static Setting<Color> passiveColour = new Setting<>("Colour", new Color(0, 255, 0, 180))
            .setDescription("The colour to render the passive tracers in")
            .setParentSetting(passive);

    public static Setting<Boolean> mobs = new Setting<>("Mobs", true)
            .setDescription("Draws lines to monsters");

    public static Setting<Color> mobColour = new Setting<>("Colour", new Color(255, 0, 0, 180))
            .setDescription("The colour to render the mob tracers in")
            .setParentSetting(mobs);

    public static Setting<Boolean> players = new Setting<>("Players", true)
            .setDescription("Draws lines to players");

    public static Setting<Color> playerColour = new Setting<>("Colour", new Color(255, 255, 255, 180))
            .setDescription("The colour to render the player tracers in")
            .setParentSetting(players);

    public static Setting<Boolean> crystals = new Setting<>("Crystals", true)
            .setDescription("Draws lines to ender crystals");

    public static Setting<Color> crystalColour = new Setting<>("Colour", new Color(200, 0, 200, 180))
            .setDescription("The colour to render the ender crystal tracers in")
            .setParentSetting(crystals);

    public static Setting<Float> lineWidth = new Setting<>("LineWidth", 0.5f, 0.1f, 2f, 0.1f)
            .setDescription("How thick to render the lines");

    public Tracers() {
        super("Tracers", Category.RENDER, "Draws lines to entities in the world");

        INSTANCE = this;
    }

    @SubscribeEvent
    public void onRenderWorld(RenderWorldLastEvent event) {
        mc.world.loadedEntityList.forEach(entity -> {
            if (EntityUtil.isEntityAllowed(entity, players.getValue(), mobs.getValue(), passive.getValue()) && entity != mc.player || entity instanceof EntityEnderCrystal && crystals.getValue()) {
                RenderUtil.drawTracer(entity, lineWidth.getValue(), getColourByEntity(entity));
            }
        });
    }

    /**
     * Gets the entity's colour
     *
     * @param entityIn The entity
     * @return The entity's colour
     */
    private Color getColourByEntity(Entity entityIn) {
        if (EntityUtil.isPassive(entityIn)) {
            return passiveColour.getValue();
        }

        if (EntityUtil.isMonster(entityIn)) {
            return mobColour.getValue();
        }

        if (entityIn instanceof EntityPlayer) {
            return playerColour.getValue();
        }

        if (entityIn instanceof EntityEnderCrystal) {
            return crystalColour.getValue();
        }

        return passiveColour.getValue();
    }

}

File: src/main/java/com/paragon/client/systems/module/impl/render/Tracers.kt
package com.paragon.client.systems.module.impl.render

import com.paragon.api.module.Category
import com.paragon.api.module.Module
import com.paragon.api.setting.Setting

import com.paragon.api.util.Wrapper
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.client.event.RenderWorldLastEvent
import com.paragon.api.util.entity.EntityUtil
import com.paragon.api.util.render.ColourUtil
import net.minecraft.entity.item.EntityEnderCrystal
import com.paragon.api.util.render.RenderUtil
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.math.MathHelper
import java.awt.Color
import java.util.function.Consumer
import kotlin.math.min

/**
 * @author Surge
 */
class Tracers : Module("Tracers", Category.RENDER, "Draws lines to entities in the world") {
    private val passive = Setting<Boolean?>("Passives", true)
        .setDescription("Draws lines to passive entities")

    private val passiveColour = Setting("Colour", Color(0, 255, 0, 180))
        .setDescription("The colour to render the passive tracers in")
        .setParentSetting(passive)
        .setVisibility { distance.value != Distance.COLOUR }

    private val mobs = Setting<Boolean?>("Mobs", true)
        .setDescription("Draws lines to monsters")

    private val mobColour = Setting("Colour", Color(255, 0, 0, 180))
        .setDescription("The colour to render the mob tracers in")
        .setParentSetting(mobs)
        .setVisibility { distance.value != Distance.COLOUR }

    private val players = Setting<Boolean?>("Players", true)
        .setDescription("Draws lines to players")

    private val playerColour = Setting("Colour", Color(255, 255, 255, 180))
        .setDescription("The colour to render the player tracers in")
        .setParentSetting(players)
        .setVisibility { distance.value != Distance.COLOUR }

    private val crystals = Setting<Boolean?>("Crystals", true)
        .setDescription("Draws lines to ender crystals")

    private val crystalColour = Setting("Colour", Color(200, 0, 200, 180))
        .setDescription("The colour to render the ender crystal tracers in")
        .setParentSetting(crystals)
        .setVisibility { distance.value != Distance.COLOUR }

    private val distance = Setting("Distance", Distance.ALPHA)
        .setDescription("Change the colour depending on the distance to the entity")

    private val distanceDivider = Setting("Divider", 1500f, 0f, 5000f, 100f)
        .setDescription("The divider to use when calculating the distance factor")
        .setParentSetting(distance)

    private val lineWidth = Setting("LineWidth", 0.5f, 0.1f, 2f, 0.1f)
        .setDescription("How thick to render the lines")

    @SubscribeEvent
    fun onRenderWorld(event: RenderWorldLastEvent?) {
        minecraft.world.loadedEntityList.forEach(Consumer { entity: Entity ->
            if (EntityUtil.isEntityAllowed(entity, players.value!!, mobs.value!!, passive.value!!) && entity !== minecraft.player || entity is EntityEnderCrystal && crystals.value!!) {
                RenderUtil.drawTracer(entity, lineWidth.value, getColourByEntity(entity))
            }
        })
    }

    /**
     * Gets the entity's colour
     *
     * @param entityIn The entity
     * @return The entity's colour
     */
    private fun getColourByEntity(entityIn: Entity): Color {
        var colour: Color = passiveColour.value

        if (EntityUtil.isPassive(entityIn)) {
            colour = passiveColour.value
        }

        if (EntityUtil.isMonster(entityIn)) {
            colour = mobColour.value
        }

        if (entityIn is EntityPlayer) {
            colour = playerColour.value
        }

        if (entityIn is EntityEnderCrystal) {
            colour = crystalColour.value
        }

        if (distance.value != Distance.OFF) {
            val factor: Float = MathHelper.clamp((minecraft.player.getDistanceSq(entityIn) / distanceDivider.value).toFloat(), 0f, 1f)

            colour = if (distance.value == Distance.COLOUR) {
                ColourUtil.integrateAlpha(Color(1f - factor, factor, 0f), colour.alpha.toFloat())
            } else {
                ColourUtil.integrateAlpha(colour, 255f * (1 - factor))
            }
        }

        return colour
    }

    enum class Distance {
        /**
         * Red -> Green depending on distance
         */
        COLOUR,

        /**
         * Transparency based on distance
         */
        ALPHA,

        /**
         * Don't do anything
         */
        OFF
    }
}